src/RawPOSIX/src/safeposix/syscalls/fs_calls.rs

```diff
@@ -780,6 +780,46 @@ impl Cage {
         
     }
 
+    /* 
+    *   dup3(): duplicates oldfd to newfd, similar to dup2, but with additional functionality
+    *   First, it allows specifying flags, such as O_CLOEXEC, to set the close-on-exec flag atomically
+    *   second, unlike dup2, dup3 prevents newfd from being the same as oldfd, returning an error (EINVAL) instead of closing newfd
+    *   Link to the manpage: https://linux.die.net/man/2/dup3
+    */
+    pub fn dup3_syscall(&self, old_virtualfd: i32, new_virtualfd: i32, flags: i32) -> i32 {
+        if old_virtualfd < 0 || new_virtualfd < 0 {
+            return syscall_error(Errno::EBADF, "dup3", "Bad File Descriptor");
+        }
+
+        //if flags is not 0 or O_CLOEXEC return error
+        if flags != 0 && flags != O_CLOEXEC{
+            return syscall_error(Errno::EINVAL, "dup3", "Invalid flags");
+        }
+
+        //newfd can not be equal to oldfd
+        if old_virtualfd == new_virtualfd {
+            return syscall_error(Errno::EINVAL, "dup3", "oldfd and newfd must be different");
+        }
+
+        match fdtables::translate_virtual_fd(self.cageid, old_virtualfd as u64) {
+            Ok(old_vfd) => {
+                let new_kernelfd = unsafe {
+                    libc::dup(old_vfd.underfd as i32)
+                };
+                let _ret_kernelfd = unsafe{ libc::dup2(old_vfd.underfd as i32, new_kernelfd) };
+
+                //right now flags can only be O_CLOEXEC or 0, by definition of O_CLOEXEC should_cleexec will be true(the value of O_CLOEXEC is a positive value)
```

Comment:
this comment probably makes sense to be with the CLOEXEC check at the beginning of the function